### PR TITLE
[FIX] web: `formView` mobile gaps

### PR DESCRIPTION
**Pull request by morm-odoo**
---
**NOTE: this PR is handled by mano-odoo**

---

Prior to this PR, there was undesired vertical gap between fields in the formView (mobile only).

Adds a custom property on the generic `.o_field_widget` component to increase flexibility on its margin-bottom property. Thus we adapt the rules which were overriding the margin-bottom property to use the new custom property instead.

The form-break-table mixin is adapted to remove the unwanted extra spacing. 
To handle the `o_address_type` and `o_address_format` exception a new boolean parameter is introduced in the mixin to differentiate editable form views: `$-is-editable`.

This PR also changes the `type` field label tag to better fit label styling.

task-3806250

image

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/views/form/form_controller.scss [hunk 1]
```diff
@@ -1117,7 +1118,13 @@
 
 .o_form_view.o_xxs_form_view {
     .o_inner_group {
-        @include form-break-table;
+        .o_wrap_field {
+            .o_cell {
+                .o_field_widget {
+                    margin-bottom: 0;
```
Comment (stefanorigano) [inline addons/web/static/src/views/form/form_controller.scss:1124]: As a general approach, we do tend to customise these properties contextually and by using CSS variables.

Depending by the amount of changes necessary, you may want to take the approach in consideration.
I'm not saying that you have to, a one-line fix would be preferred indeed, I'm just exposing some possibilities 😉

More info:
https://www.odoo.com/documentation/master/contributing/development/coding_guidelines.html#css-variables

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Hello @morm-odoo , thanks for your work 👍
Some remarks for you in the code.

note: no need of adding the `[DRAFT]` prefix title in the community repo, just create these PRs as a draft, as you correctly did 😉

File: addons/web/static/src/views/form/form_controller.scss [hunk 2]
```diff
@@ -1119,6 +1120,13 @@
 .o_form_view.o_xxs_form_view {
     .o_inner_group {
         @include form-break-table;
+        .o_wrap_field {
+            .o_cell {
+                .o_field_widget {
+                    margin-bottom: 0;
```
Comment (stefanorigano) [inline addons/web/static/src/views/form/form_controller.scss:1126]: Thanks for you work @morm-odoo 👍 

Unfortunately this rule will be triggered for address fields when in edit-mode as well, while it should **not**.
Could you address it? Thanks! 

| . | Edit Mode | Read-only |
|--------|--------|--------|
| master | image | image |
| this PR | image | image |

**Review by stefanorigano — CHANGES_REQUESTED** [review]

File: addons/web/static/src/views/form/form_controller.scss (hunk 2 above)
Comment (morm-odoo) [reply addons/web/static/src/views/form/form_controller.scss:1126]: @stefanorigano Thanks for the feedback! 
I fixed this issue with adding `&:not(.o_address_format)`, this should avoid the problem in all the address fields.

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Hello @morm-odoo 

> I fixed this issue with adding &:not(.o_address_format), this should avoid the problem in all the address fields.

It avoids the address fields problem, yes... but just because your code it's _not_ triggered at all, neither where it was supposed to.
Please remember to test your code (... and to request a review if you need one).

Thanks

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Hello @morm-odoo , thanks for your work.
Beside the remarks in the code , we should retain a `margin-bottom`  for the `<select>`  in the address field, see following screenshot.
Thanks!

| this PR | expected result |
|--------|--------|
| image | image|

**Comment by morm-odoo** [comment]
@stefanorigano Thanks a lot for your review :) 
I thought about two solutions for the `margin-bottom` of the `selection field` for the address.
I can either modify directly the `xml field` (inside `res_partner_views.xml`) to add the `o_address_format` class, and It would get the correct variable, or I could target the `[name="address_name"]` in scss and add the variable with the correct margin. 
If you think one of those solutions is good, which one would you think is best?

**Comment by stefanorigano** [comment]
> I thought about two solutions for the margin-bottom of the selection field for the address.
I can either modify directly the xml field (inside res_partner_views.xml) to add the o_address_format class, and It would get the correct variable, or I could target the [name="address_name"] in scss and add the variable with the correct margin.
If you think one of those solutions is good, which one would you think is best?

@morm-odoo , I tested your branch in order to provide you insights on which approach to use, as requested...  but apparently you have chosen and pushed an approach already, so I wasted my time testing in a faulty environment, ok 😅
Let's just request a review when you think that your code is ready, ok?

Thanks

**Comment by morm-odoo** [comment]
> > I thought about two solutions for the margin-bottom of the selection field for the address.
> > I can either modify directly the xml field (inside res_partner_views.xml) to add the o_address_format class, and It would get the correct variable, or I could target the [name="address_name"] in scss and add the variable with the correct margin.
> > If you think one of those solutions is good, which one would you think is best?
> 
> @morm-odoo , I tested your branch in order to provide you insights on which approach to use, as requested... but apparently you have chosen and pushed an approach already, so I wasted my time testing in a faulty environment, ok 😅 Let's just request a review when you think that your code is ready, ok?
> 
> Thanks

@stefanorigano I'm sorry, my message was not clear. I indeed pushed one of the solution I proposed, I should have said that, and that I was only suggesting (and asking advice) on another solution possible.

**Comment by mano-odoo** [comment]
Note: 
- CSS variable behavior changed, 0 by default, but `2 * $o-form-spacing-unit` in case of a field_widget inside `o_address_format`.
- row-gap: 0; is applied in the mixin (it was impacting other viewports than mobile)

For the select field, this one is actually quite tricky. It's either a label, or a field which changes the address_type depending on the is_company variable. To avoid applying unrelated or unwanted styling change to such a "specific" field I created a class to target it directly. It feels safer than applying the o_address_format class (which is made for the content that are under this field). 

I think there is room for improvement on this field. It's essentially the "parent" of the inner address related field. But it doesn't look nor behave as such, because currently modifying one type of address is modifying all the other type of addresses for a contact🤔 

I already improved it a bit but it's maybe worth creating a task and investigating the expected and wanted behavior of this field/label, what do you think? @stefanorigano, as for me I find it confusing.

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Hello @mano-odoo , thanks.
A couple of remarks for you.

Btw, either we modify the mixin or we add an hook at the main component, I'm ok with both approaches, yes. Personally I do prefer the latter, but since it's the mixin behaves like that already, I'm ok with your proposal.

> it's maybe worth creating a task and investigating the expected and wanted behavior of this field/label, what do you think? 

Not sure what 's the problem you want to solve here, let's discuss it tomorrow ;)

File: addons/web/static/src/views/form/form_controller.scss [hunk 3]
```diff
@@ -26,7 +27,7 @@
             }
 
             .o_field_widget {
-                margin-bottom: $o-form-spacing-unit * 2;
+                margin-bottom: var(--fieldWidget-xxs-margin-bottom, 0);
```
[the remark this answers is not in the document]
Comment (mano-odoo) [reply addons/web/static/src/views/form/form_controller.scss:30]: I did a push which renames the variable to reflect its scope. It's a bit longer but at least it achieves the goal of describing its scope.

However the comment describing the mixin is confusing. As you've said even if the mixin is called in modal-sm (300px width) and in xxs form views, it can be called elsewhere in the future.
 
Therefore `// XXS form view specific rules` doesn't describe well what the mixin is doing. Should we reword it to something such as `// Form view rules to break the default table layout`. It's a bit further from the scope of this task but it's a good opportunity to do so, what do you think?

**Comment by mano-odoo** [comment]
In the end I decided to go for the variable inside the mixin because the approach we discussed together was not providing the expected behavior. 

Using `.o_form_editable & {` inside the mixin was not an option as the nesting above the mixin call was effectively creating the following selector: `& .o_form_editable .modal-sm .o_form_view .o_inner_group {`
Instead of the desired `& .modal-sm .o_form_view .o_form_editable .o_inner_group {`

We still have the option to include the .o_inner_group class inside the mixin but that would make the mixin very specific. So I didn't go for this option, let me know what you think @stefanorigano

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Almost there, thanks @mano-odoo 👍
(🎂)

**Comment by mano-odoo** [comment]
- 1st push is diff
- 2nd push is rebase
- 3rd push commit message variable change(forgot)

File: addons/web/static/src/views/fields/fields.scss [hunk 4]
```diff
@@ -93,6 +93,7 @@
     // Default display and alignment of widget and internal <input/>
     text-align: inherit;
     display: var(--fieldWidget-display, inline-block);
+    margin-bottom: var(--fieldWidget-margin-bottom, initial);
```
Comment (stefanorigano) [inline addons/web/static/src/views/fields/fields.scss:96]: `initial`: why this fallback value it's needed? 🤔

File: addons/web/static/src/views/form/form_controller.scss [hunk 5]
```diff
@@ -12,44 +12,45 @@
...
+    --fieldWidget-margin-bottom: 0;
+
     grid-template-columns: 1fr;
     margin-bottom: $o-form-spacing-unit * 4;
+    row-gap: 0;
 
     .o_wrap_field {
         .o_cell {
             max-width: 100%;
-
```
Comment (stefanorigano) [inline addons/web/static/src/views/form/form_controller.scss:23]: Why? Let's keep the empty line ;)

File: addons/web/static/src/views/form/form_controller.scss [hunk 6]
```diff
@@ -12,44 +12,45 @@
...
             &.o_wrap_label {
                 line-height: $o-label-font-size-factor;
             }
-
```
Comment (stefanorigano) [inline addons/web/static/src/views/form/form_controller.scss:27]: Let's keep the empty line, here as well ;)

File: addons/web/static/src/views/form/form_controller.scss [hunk 7]
```diff
@@ -12,44 +12,45 @@
...
-                margin-bottom: $o-form-spacing-unit * 2;
-
                 &:not(.o_field_boolean){
                     width: 100%;
                 }
-
-                > .o_field_widget {
-                    margin-bottom: 0;
-                }
-
```
Comment (stefanorigano) [inline addons/web/static/src/views/form/form_controller.scss:38]: Please let's keep an empty separation line between selectors 👍

File: addons/web/static/src/views/form/form_controller.scss [hunk 8]
```diff
@@ -12,44 +12,45 @@
...
                 &.o_field_boolean {
                     margin-right: 0;
                 }
             }
-
```
Comment (stefanorigano) [inline addons/web/static/src/views/form/form_controller.scss:43]: Why ? :)

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Hello @mano-odoo , one question and some little remarks for you, thanks 👍

File: addons/web/static/src/views/fields/fields.scss (hunk 4 above)
Comment (mano-odoo) [reply addons/web/static/src/views/fields/fields.scss:96]: It's absolutely useless sorry about that 😅 If no margin is set it would be initial value anyways, if a margin-bottom is set it would override the property, and if the custom property is set it would take the CPP value.

**Review by stefanorigano — APPROVED** [review]
LGTM, thanks @mano-odoo 👍

**Comment by chgo-odoo** [comment]
Hello @odoo/rd-framework-js 👋

I guess this one is part of your scope as it's about the form view + it's mobile related. Could you have a look at it, please ?

Thanks !

File: odoo/addons/base/views/res_partner_views.xml [hunk 9]
```diff
@@ -174,9 +174,9 @@
 
                     <group>
                         <group>
-                            <span class="o_form_label o_td_label" name="address_name">
+                            <span class="o_form_label o_td_label o_address_type" name="address_name">
```
Comment (aab-odoo) [inline odoo/addons/base/views/res_partner_views.xml:177]: can we get rid of this new classname?

Comment (stefanorigano) [reply odoo/addons/base/views/res_partner_views.xml:177]: Not really @aab-odoo , we need to bind the css to this element somehow.
Since its siblings comes with the `o_address_format ` class already, it make sense to me add this new class.
(fyi: mano is currently  off, I'm handling this)

**Review by aab-odoo — APPROVED** [review]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo rebase-ff
